Add BDA resampling methods

plaplant opened the pull request:
## Description
This PR adds methods on `UVData` objects for resampling datasets to either a uniform shorter integration time (`bda_upsample`) or longer time (`bda_downsample`). There are other related smaller changes, such as adding a `copy` method to the `UVData` object as well as support for phasing on metadata-only objects.

## Motivation and Context
With the advent of BDA datasets, resampling to a common integration time is important to test resampling datasets and understanding how code will be affected. It is also useful for reimplementing existing applications after a dataset has been resampled.

Closes #428.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change


## Checklist:
<!--- You may remove the checklists that don't apply to your change type(s) or just leave them empty -->
<!--- Go over all the following points, and replace the space with an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.

New feature checklist:
- [x] I have added or updated the docstrings associated with my feature using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to highlight my new feature (if appropriate).
- [x] I have added tests to cover my new feature.
- [x] All new and existing tests pass in both python 2 and python 3.
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

jsdillon commented:
Sorry to take a while to look at this @plaplant (I've been traveling) but this looks pretty good from a high-level interface perspective. I think this will do what we need on the `hera_cal` side, but I won't be 100% sure until I go ahead and implement BDA support.

Two minor comments:
- It would be nice to have a `bda_resample()` function that dynamically upsamples or downsamples baselines as appropriate.
- The documentation for ordering parameters should refer to the documentation on `reorder_blts`

bhazelton commented:
Might need an option to do integrating vs averaging -- to match uncalibrated HERA data.

bhazelton commented:
@adampbeardsley @mkolopanis I think this branch might be ready for another look. I think we've addressed all the issues you raised before and the others that came up since. The only open thing I'm aware of is whether we want to rename the `integrate` keyword.

bhazelton commented:
Need to add something to the history to mark that resampling happened.

bhazelton commented:
change `integrate` to `summing_correlator_mode`

bhazelton commented:
change `bda_*` to `*_in_time`

mkolopanis commented on pyuvdata/uvdata.py:
```diff
@@ -4628,3 +4654,497 @@ def inflate_by_redundancy(self, tol=1.0, blt_order='time', blt_minor_order=None)
+                        if (temp_flag[temp_idx]).any():
+                            ax1_inds, ax2_inds, ax3_inds = np.nonzero(temp_flag[temp_idx])
+                            for ii in range(ax1_inds.size):
+                                mask[:, ax1_inds[ii], ax2_inds[ii], ax3_inds[ii]] = False
```
I'm a little confused at the need to iterate through the arrays here. Is it not able to just be indexed as `mask[:, ax1_inds, ax2_inds, ax3_inds]`?  We're saying here that all indices for the 0th axis should be flagged right?

adampbeardsley replied:
I could be wrong, but I don't think that's quite the same. For example, if `ax1_inds = [0, 1, 2]`, `ax2_inds = [0, 1, 2]`, and `ax3_inds = [0, 1, 2]`, the code as written will set the diagonal to `False`, whereas your code would set the whole thing to `False`. In other words, code as written allows for fine grained selection of each individual pixel, but yours will only unflag slices.

mkolopanis replied:
we should test this. If we don't need a for loop i would prefer to avoid it.

plaplant replied:
I've tested this empirically. Dropping in the `ax?_inds` seems to Do The Right Thing and just zip them together (rather than take the outer product). I've opted for @mkolopanis's approach to eliminate the for loop.

mkolopanis commented on pyuvdata/uvdata.py:
```diff
@@ -506,6 +507,28 @@ def check(self, check_extra=True, run_check_acceptability=True):
+        uv = UVData()
+        for param in self:
+            # parameter names have a leading underscore we want to ignore
+            if metadata_only and param[1:] in self._data_params:
```
also worth noting this would remove all leading `_` characters.

plaplant replied:
Good suggestion, I used `lstrip`.

adampbeardsley commented on pyuvdata/uvdata.py:
```diff
@@ -4628,3 +4654,497 @@ def inflate_by_redundancy(self, tol=1.0, blt_order='time', blt_minor_order=None)
+                                  "expected".format(bl=self.baseline_to_antnums(bl)))
+
+            else:
+                # varying integration times for this baseline, need to be more careful
```
I _think_ one could simply run this else block instead of having the if statement at all.

plaplant replied:
I think you're right, but I've kept it as is because I'm lazy ;-)

adampbeardsley commented on pyuvdata/uvdata.py:
```diff
@@ -4628,3 +4654,497 @@ def inflate_by_redundancy(self, tol=1.0, blt_order='time', blt_minor_order=None)
+                    temp_baseline[temp_idx] = bl
+                    # this might be wrong if some of the constituent times are *totally* flagged
+                    averaging_idx = bl_inds[summing_idx:summing_idx + n_sum]
+                    temp_time[temp_idx] = np.average(self.time_array[averaging_idx])
```
I think there's an implicit assumption here (and line 5006) that the integration times are uniform within this baseline. Here I think it can be fixed with something like:
```
temp_time[temp_idx] = (np.sum(self.time_array[averaging_idx] * self.integration_integration_time[averaging_idx])
                       / np.sum(self.integration_integration_time[averaging_idx])
```
And a similar weighted mean would make sense below.

plaplant replied:
Good point! I've updated it as indicated here and in the actual data averaging below (which was slightly trickier to get right because of the mis-match in array dimensions between the data and `integration_time`).

adampbeardsley commented on pyuvdata/uvdata.py:
```diff
@@ -4628,3 +4654,497 @@ def inflate_by_redundancy(self, tol=1.0, blt_order='time', blt_minor_order=None)
+
+        # drop data where we downsampled
+        inds_to_keep = np.nonzero(self.integration_time >= min_int_time)
+        self.baseline_array = self.baseline_array[inds_to_keep]
```
Basically from here to the end of the function is repeated between up/down sampling. Would it make sense to abstract this out to a function so we don't have so much repeated code?

plaplant replied:
Yeah, probably better to have less repeated code. I've pulled out most of the repeated bits into a common function.

plaplant commented on pyuvdata/uvdata.py:
```diff
@@ -4628,3 +4654,497 @@ def inflate_by_redundancy(self, tol=1.0, blt_order='time', blt_minor_order=None)
+        self.check()
+
+        # add to the history
+        history_update_string = (" Upsampled data to {:f} second integration time "
```
Fixed, thanks!

plaplant commented on pyuvdata/uvdata.py:
```diff
@@ -4628,3 +4654,497 @@ def inflate_by_redundancy(self, tol=1.0, blt_order='time', blt_minor_order=None)
+            for ii, idx in enumerate(range(i0, i1)):
+                idx2 = ii + offset + n2 - n_new_samples[i]
+                nt = ((t0 * units.day) + (dt * idx2 * units.s)).to(units.day).value
+                temp_time[idx] = nt
```
I've added some comments. Hopefully it's clearer now.

plaplant commented on pyuvdata/uvdata.py:
```diff
@@ -4628,3 +4654,497 @@ def inflate_by_redundancy(self, tol=1.0, blt_order='time', blt_minor_order=None)
+        Parameters
+        ----------
+        max_int_time : float
+            Maximum integration time to upsample the to in seconds.
```
Thanks!

plaplant commented:
@adampbeardsley @mkolopanis thanks for the (multiple) thorough reviews! I think I've addressed all of the comments, but let me know if there is anything else that needs attention.

mkolopanis commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -3909,3 +3914,1014 @@ def test_lsts_from_time_with_only_unique():
+    out_ns = uv_object.get_nsamples(0, 1)
+    assert np.isclose(init_ns[0, 0, 0], out_ns[0, 0, 0])
+
+    # try again with allow_drift=False
```
might even break up the with and without `allow_drift` case

plaplant replied:
Done.

mkolopanis commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -3909,3 +3914,1014 @@ def test_lsts_from_time_with_only_unique():
+    # make copies for later
+    uv_object2 = uv_object.copy()
+    uv_object3 = uv_object.copy()
+    uv_object4 = uv_object.copy()
```
again these sound like multiple tests

plaplant replied:
Indeed, these are now separate.

mkolopanis commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -3909,3 +3914,1014 @@ def test_lsts_from_time_with_only_unique():
+    assert np.isclose(init_data_136_137[0, 0, 0], out_data_136_137[0, 0, 0])
+
+    # again, with only_downsample set
+    uv_object2.resample_in_time(8, only_downsample=True)
```
The only this branch or only that branch of the `resample_in_time` tests could probably be their own tests too.

plaplant replied:
Done.

plaplant commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -3495,17 +3500,17 @@ def test_set_uvws_from_antenna_pos():
         [True, 'gcrs', 'gcrs'],
-        message='Warning: Data will be unphased'
+        message='Data will be unphased'
     )
```
Thanks for catching this, I've explicitly tested the error messages are what we expect.

plaplant commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -3909,3 +3914,1014 @@ def test_lsts_from_time_with_only_unique():
+
+    # this should be true because there are no flags
+    out_ns = uv_object_copy2.get_nsamples(0, 1)
+    assert np.isclose(init_ns[0, 0, 0], out_ns[0, 0, 0])
```
Done, these are now separate tests.

plaplant commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -3909,3 +3914,1014 @@ def test_lsts_from_time_with_only_unique():
+    out_ns = uv_object_copy2.get_nsamples(0, 1)
+    assert np.isclose(init_ns[0, 0, 0], out_ns[0, 0, 0])
+
+    return
```
Done.

plaplant commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -3909,3 +3914,1014 @@ def test_lsts_from_time_with_only_unique():
+                         message="All values in integration_time array are already shorter")
+    assert uv_object == uv_object2
+
+    return
```
Done.

plaplant commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -3909,3 +3914,1014 @@ def test_lsts_from_time_with_only_unique():
+    """Test the upsample_in_time method"""
+    uv_object = UVData()
+    testfile = os.path.join(DATA_PATH, 'zen.2458661.23480.HH.uvh5')
+    uv_object.read(testfile)
```
Good point, I've added a fixture for this test file (and the BDA one, because I broke the test that needed it up into separate tests.

plaplant commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -3909,3 +3914,1014 @@ def test_lsts_from_time_with_only_unique():
+    uv_object_copy = uv_object.copy()
+    uv_object_copy2 = uv_object.copy()
+    uv_object_copy3 = uv_object.copy()
+    uv_object_copy4 = uv_object.copy()
```
Done.

plaplant commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -3909,3 +3914,1014 @@ def test_lsts_from_time_with_only_unique():
+    uv_object_copy = uv_object.copy()
+    uv_object_copy2 = uv_object.copy()
+    uv_object_copy3 = uv_object.copy()
+    uv_object_copy4 = uv_object.copy()
```
Done.

plaplant commented on pyuvdata/tests/test_uvdata.py:
```diff
@@ -3909,3 +3914,1014 @@ def test_lsts_from_time_with_only_unique():
+    # check that there are no flags
+    assert np.nonzero(uv_object.flag_array)[0].size == 0
+
+    # try again with allow_drift=False
```
Done.
